src/perl/bin/sample_udf_updater

```diff
@@ -0,0 +1,240 @@
+use Carp;
+use Readonly;
+use XML::LibXML;
+use List::MoreUtils qw/uniq/;
+
+use wtsi_clarity::clarity::process;
+
+with qw/wtsi_clarity::util::configurable
+        wtsi_clarity::util::roles::clarity_request
+        wtsi_clarity::util::clarity_elements/;
+
+our $VERSION = '0.0';
+
+## no critic(ValuesAndExpressions::RequireInterpolationOfMetachars)
+Readonly::Scalar my $SAMPLE_URI_PATH  => q{/art:details/art:artifact/sample/@uri};
+## use critic
+Readonly::Scalar my $SAMPLE_PATH      => q{/smp:details/smp:sample};
+
+has 'process_url'  => (
+  isa             => 'Str',
+  is              => 'ro',
+  required        => 1,
+);
+
+has 'process_doc'  => (
+  isa             => 'wtsi_clarity::clarity::process',
+  is              => 'ro',
+  required        => 0,
+  traits          => [ 'NoGetopt' ],
+  lazy_build      => 1,
+  handles         => {'findnodes' => 'findnodes'},
+);
+sub _build_process_doc {
+  my ($self) = @_;
+  my $xml = $self->fetch_and_parse($self->process_url);
+  return wtsi_clarity::clarity::process->new(xml => $xml, parent => $self);
+}
+
+has 'udf_name'  => (
+  isa             => 'Str',
+  is              => 'ro',
+  required        => 1,
+);
+
+has 'udf_value'  => (
+  isa             => 'Str',
+  is              => 'ro',
+  required        => 1,
+);
+
+has '_input_artifacts' => (
+  isa             => 'XML::LibXML::Document',
+  is              => 'rw',
+  required        => 0,
+  lazy_build      => 1,
+);
+sub _build__input_artifacts {
+  my $self = shift;
+
+  return $self->process_doc->input_artifacts;
+}
+
+has '_samples' => (
+  isa             => 'XML::LibXML::Document',
+  is              => 'rw',
+  required        => 0,
+  lazy_build      => 1,
+);
+sub _build__samples {
+  my $self = shift;
+
+  my @sample_uri_node_list = $self->_input_artifacts->findnodes($SAMPLE_URI_PATH)->get_nodelist();
+  my @sample_uris = map { $_->getValue } @sample_uri_node_list;
+
+  return $self->request->batch_retrieve('samples', \@sample_uris);
+}
+
+sub _update_samples_with_udfs {
+  my $self = shift;
+
+  my @sample_node_list = $self->_samples->findnodes($SAMPLE_PATH)->get_nodelist();
+
+  foreach my $sample (@sample_node_list) {
+    $self->_add_udf_to_sample($sample);
+  }
+
+  return;
+}
+
+sub _add_udf_to_sample {
+  my ($self, $sample_xml) = @_;
+
+  my $new_udf_element = XML::LibXML::Element->new('field');
+  $new_udf_element->setNamespace("http://genologics.com/ri/userdefined", "udf");
+  $new_udf_element->setAttribute('type', 'String');
+  $new_udf_element->setAttribute('name', $self->udf_name);
+  $new_udf_element->appendTextNode($self->udf_value);
+  $sample_xml->addChild($new_udf_element);
+
+  return;
+}
+
+sub _save_updated_samples {
+  my ($self) = @_;
+
+  $self->request->batch_update('samples', $self->_samples);
+
+  return;
+}
+
+sub run {
+  my $self = shift;
+
+  $self->_update_samples_with_udfs();
+
+  $self->_save_updated_samples();
+
+  return;
+}
+
+my $updater = sample_udf_updater->new_with_options();
```

Comment:
Yeah, I think change it to just `new`. The reason `new_with_options` is working is because it's used in the `clarity_request` role, so if that role got rid of `MooseX::Getopt` (which it probably should!) this would stop working.